### PR TITLE
Set error to waitExpr

### LINKED ISSUE #29844: Wait action should return an error (wait failure) if ask to wait on a completed future
**Description:**
$title.

Spec issue:https://github.com/ballerina-platform/ballerina-spec/issues/802

**Steps to reproduce:**

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by KRVPerera**
Fixes #29844

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-debugger-integration-test/src/test/java/org/ballerinalang/debugger/test/adapter/CallStackDebugTest.java [hunk 1]
```diff
@@ -65,7 +65,7 @@ public void stackFrameDebugTest() throws BallerinaTestException {
         debugTestRunner.assertCallStack(frames[2], "func2", 23, "main.bal");
         debugTestRunner.assertCallStack(frames[3], "func1", 19, "main.bal");
         debugTestRunner.assertCallStack(frames[4], "addition", 14, "main.bal");
-        debugTestRunner.assertCallStack(frames[5], "start:f1", 2, "main.bal");
+        debugTestRunner.assertCallStack(frames[5], "start:anonymous", 2, "main.bal");
```
Comment (KRVPerera) [inline tests/jballerina-debugger-integration-test/src/test/java/org/ballerinalang/debugger/test/adapter/CallStackDebugTest.java:68]: Needs to fix this

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 2]
```diff
@@ -3700,8 +3703,143 @@
         }
         BFutureType futureType = new BFutureType(TypeTags.FUTURE, type, null);
         checkExpr(expr, env, futureType);
+        setEventualTypeForExpression(expr, type);
     }
 
+    // eventual type if not directly referring a worker is T|error. future<T> --> T|error
+    private void setEventualTypeForExpression(BLangExpression expression,
+                                              BType currentExpectedType) {
+        if ((expression == null) || (expression.getKind() != NodeKind.SIMPLE_VARIABLE_REF)) {
```
Comment (rdhananjaya) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:3712]: I don't think we can limit this for simple_variable_references, future could be coming from a map, array, from a invocation etc. 
```ballerina
function bar() returns future<int> {
    function () returns int f = () => 0;
    return start f();
}

public function main() {
    int x = wait bar();
}
```

Comment (rdhananjaya) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:3712]: Oh my bad, I was looking at wrong function. Please disregard.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 3]
```diff
@@ -3743,6 +3881,12 @@
             // Handle other types except for semantic errors
             resultType = ((BFutureType) resultType).constraint;
         }
+
+        if (waitExpr.getExpression().getKind() == NodeKind.SIMPLE_VARIABLE_REF) {
```
Comment (rdhananjaya) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:3885]: regular wait is when size of getExpression is one.
We need to support non simple_var_ref situations.

```ballerina
// Coppied from other comment
function bar() returns future<int> {
    function () returns int f = () => 0;
    return start f();
}

public function main() {
    int x = wait bar();
}
```

**Review by rdhananjaya — APPROVED** [review]

**Comment by rdhananjaya** [comment]
Let's fix the review suggestions in a separate PR as they are more of improvements

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 4]
```diff
@@ -263,6 +263,7 @@
     private Stack<SymbolEnv> queryEnvs, prevEnvs;
     private Stack<BLangSelectClause> selectClauses;
     private BLangMissingNodesHelper missingNodesHelper;
+    private String nonWorkerReferenceName;
```
Comment (rdhananjaya) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:266]: Shall we look into getting rid of this field, looks like the function calls doesn't go that deep, may be we can use empty/null to differentiate

**Comment by KRVPerera** [comment]
Fixing feedback in PR
https://github.com/ballerina-platform/ballerina-lang/pull/30631

File: tests/jballerina-debugger-integration-test/src/test/java/org/ballerinalang/debugger/test/adapter/CallStackDebugTest.java (hunk 1 above)
Comment (KRVPerera) [reply tests/jballerina-debugger-integration-test/src/test/java/org/ballerinalang/debugger/test/adapter/CallStackDebugTest.java:68]: Created https://github.com/ballerina-platform/ballerina-lang/issues/30636 to track this

**Pull request merged**